cpp/src/arrow/compute/kernels/take-internal.h

```diff
@@ -455,20 +493,181 @@ class TakerImpl<IndexSequence, StructType> : public Taker<IndexSequence> {
   std::vector<std::unique_ptr<Taker<IndexSequence>>> children_;
 };
 
+template <typename IndexSequence>
+class TakerImpl<IndexSequence, UnionType> : public Taker<IndexSequence> {
+ public:
+  using Taker<IndexSequence>::Taker;
+
+  Status Init() override {
+    union_type_ = checked_cast<const UnionType*>(this->type_.get());
+
+    if (union_type_->mode() == UnionMode::SPARSE) {
+      sparse_children_.resize(this->type_->num_children());
+    } else {
+      dense_children_.resize(this->type_->num_children());
+      child_length_.resize(union_type_->max_type_code() + 1);
+    }
+
+    for (int i = 0; i < this->type_->num_children(); ++i) {
+      if (union_type_->mode() == UnionMode::SPARSE) {
+        RETURN_NOT_OK(Taker<IndexSequence>::Make(this->type_->child(i)->type(),
+                                                 &sparse_children_[i]));
+      } else {
+        RETURN_NOT_OK(Taker<ArrayIndexSequence<Int32Type>>::Make(
+            this->type_->child(i)->type(), &dense_children_[i]));
+      }
+    }
+
+    return Status::OK();
+  }
+
+  Status SetContext(FunctionContext* ctx) override {
+    pool_ = ctx->memory_pool();
+    null_bitmap_builder_.reset(new TypedBufferBuilder<bool>(pool_));
+    type_id_builder_.reset(new TypedBufferBuilder<int8_t>(pool_));
+
+    if (union_type_->mode() == UnionMode::DENSE) {
+      offset_builder_.reset(new TypedBufferBuilder<int32_t>(pool_));
+      std::fill(child_length_.begin(), child_length_.end(), 0);
+    }
+
+    for (int i = 0; i < this->type_->num_children(); ++i) {
+      if (union_type_->mode() == UnionMode::SPARSE) {
+        RETURN_NOT_OK(sparse_children_[i]->SetContext(ctx));
+      } else {
+        RETURN_NOT_OK(dense_children_[i]->SetContext(ctx));
+      }
+    }
+
+    return Status::OK();
+  }
+
+  Status Take(const Array& values, IndexSequence indices) override {
+    DCHECK(this->type_->Equals(values.type()));
+    const auto& union_array = checked_cast<const UnionArray&>(values);
+    auto type_ids = union_array.raw_type_ids();
+
+    std::vector<uint32_t> child_counts(union_type_->max_type_code() + 1);
```

Comment:
`child_counts` is only used for dense arrays, right?

Reply:
True, and I can move this block into the union mode branches and avoid accumulating child_counts in the sparse case. I was trying to minimize code duplication and `child_counts` seemed cheap